pytest_reqs.py

```diff
@@ -46,20 +52,37 @@ def pytest_sessionstart(session):
 def pytest_collection_modifyitems(config, session, items):
     if config.option.reqs:
         check_requirements(config, session, items)
+    if config.option.reqs_outdated:
+        check_outdated_requirements(config, session, items)
 
 
-def check_requirements(config, session, items):
+def get_reqs_filenames(config):
     patterns = config.patterns or DEFAULT_PATTERNS
-    filenames = set(chain.from_iterable(map(glob, patterns)))
+    return set(chain.from_iterable(map(glob, patterns)))
 
+
+def check_requirements(config, session, items):
     installed_distributions = dict(
         (d.project_name.lower(), d)
         for d in get_installed_distributions()
     )
 
     items.extend(
         ReqsItem(filename, installed_distributions, config, session)
-        for filename in filenames
+        for filename in get_reqs_filenames(config)
+    )
+
+
+def check_outdated_requirements(config, session, items):
+    pip_outdated_dists_output = check_output(
+        [executable, '-m', 'pip', 'list', '--outdated'], stderr=STDOUT
+    )
```

Comment:
Thanks for doing this the right way (i.e. by calling pip rather than using it's internal API)! I think we might want to tweak this a bit but I'm curious what your thoughts are.

I think if we added `--format=json` to this call we might have something slightly more useful to work with. For example, this is the existing error message when a requirement is at the wrong version:

> Distribution "flake8" requires flake8==3.0.4 (from -r requirements/test.txt (line 1)) but 3.5.0 is installed

Note how it tells the user a) the full path to the requirements file it's coming from b) an exact version specifier (`flake8==3.0.4`) that it's expecting.

This is the current output when a dependency is not at the latest release:

> Distribution "flake8-import-order" is outdated (flake8-import-order (0.10) - Latest:
0.16 [wheel])

Note that this doesn't specify the original filename for the requirements file, and doesn't give a version pin. It also has the filetype for the dependency (`wheel`) which is confusing even for me, because I thought it was a mistake and was referring to the version of `wheel` that was installed.

I'd really like to see this error message become something like:

> Distribution "flake8-import-order" is outdated (from -r requirements/test.txt (line 1)), latest version is flake8-import-order==0.16

which I think will be easy to do if we switch to the JSON output.

Reply:
Yes, you are right, using the format option gives more useful information, but it's available only in the latest versions of pip (9.0.0 and above). I fixed the pr and updated the readme.